File: src/demandas/demandas.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  Post,
  Put,
  Patch,
  Param,
  Query,
  UseGuards,
  Req,
  Res,
  UseInterceptors,
  UploadedFile,
} from '@nestjs/common';
import { Response } from 'express';
import { FileInterceptor } from '@nestjs/platform-express';
import { DemandasService } from './demandas.service';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';
import { CreateDemandaDto } from './dto/create-demanda.dto';
import { UpdateDemandaDto } from './dto/update-demanda.dto';
import { ListDemandasFiltersDto } from './dto/list-demandas-filters.dto';
import { CreateDemandaFromTemplateDto } from '../templates/dto/create-demanda-from-template.dto';
import { BuscarIaDto } from './dto/buscar-ia.dto';
import { AdminGuard } from '../auth/admin.guard';
import { stringToBool } from './utils';

@Controller('demandas')
@UseGuards(JwtAuthGuard)
export class DemandasController {
  constructor(private demandasService: DemandasService) {}

  @Post()
  create(@Req() req: { user: { id: string } }, @Body() dto: CreateDemandaDto) {
    return this.demandasService.create(req.user.id, dto);
  }

  @Post('from-template/:templateId')
  createFromTemplate(
    @Req() req: { user: { id: string } },
    @Param('templateId') templateId: string,
    @Body() dto: CreateDemandaFromTemplateDto,
  ) {
    return this.demandasService.createFromTemplate(req.user.id, templateId, dto);
  }

  @Post('buscar-ia')
  async buscarIa(
    @Req() req: { user: { id: string } },
    @Body() dto: BuscarIaDto,
  ) {
    return this.demandasService.buscarIa(req.user.id, dto.query, {
      scope: dto.scope,
      context: dto.context,
    });
  }

  @Get('dashboard-kpis')
  @UseGuards(AdminGuard)
  async getDashboardKpis(@Req() req: { user: { id: string } }, @Query('avaliar_ia') avaliarIa?: string) {
    return this.demandasService.getDashboardKpis(req.user.id, avaliarIa === 'true');
  }

  @Get()
  async list(
    @Req() req: { user: { id: string } },
    @Query() query: ListDemandasFiltersDto,
  ) {
    const filters = {
      ...query,
      prioridade: stringToBool(query.prioridade),
      ocultarStandby: stringToBool(query.ocultarStandby),
      responsavelApenasPrincipal: stringToBool(query.responsavelApenasPrincipal),
    } as ListDemandasFiltersDto;
    return this.demandasService.list(req.user.id, filters);
  }

  @Get('export/excel')
  async exportExcel(
    @Req() req: { user: { id: string } },
    @Res() res: Response,
    @Query() query: ListDemandasFiltersDto,
  ) {
    const filters = {
      ...query,
      prioridade: stringToBool(query.prioridade),
      ocultarStandby: stringToBool(query.ocultarStandby),
      responsavelApenasPrincipal: stringToBool(query.responsavelApenasPrincipal),
    } as ListDemandasFiltersDto;
    const data = await this.demandasService.exportExcel(req.user.id, filters);
    res.setHeader('Content-Type', 'application/json');
    res.setHeader('Content-Disposition', 'attachment; filename=demandas.json');
    return res.send(JSON.stringify(data, null, 2));
  }

  @Get(':id')
  findOne(@Req() req: { user: { id: string } }, @Param('id') id: string) {
    return this.demandasService.findOne(req.user.id, id);
  }

  @Put(':id')
  update(
    @Req() req: { user: { id: string } },
    @Param('id') id: string,
    @Body() dto: UpdateDemandaDto,
  ) {
    return this.demandasService.update(req.user.id, id, dto);
  }

  @Delete(':id')
  @UseGuards(AdminGuard)
  remove(@Req() req: { user: { id: string } }, @Param('id') id: string) {
    return this.demandasService.remove(req.user.id, id);
  }

  @Post(':id/observacoes')
  addObservacao(
    @Req() req: { user: { id: string } },
    @Param('id') id: string,
    @Body() body: { texto: string },
  ) {
    return this.demandasService.addObservacao(req.user.id, id, body.texto);
  }

  @Patch(':id/observacoes/:observacaoId')
  updateObservacao(
    @Req() req: { user: { id: string } },
    @Param('id') id: string,
    @Param('observacaoId') observacaoId: string,
    @Body() body: { texto: string },
  ) {
    return this.demandasService.updateObservacao(req.user.id, id, observacaoId, body.texto);
  }

  @Post(':id/anexos')
  @UseInterceptors(FileInterceptor('file'))
  addAnexo(
    @Req() req: { user: { id: string } },
    @Param('id') id: string,
    @UploadedFile() file: Express.Multer.File,
    @Body() body: { nome?: string },
  ) {
    if (!file) throw new BadRequestException('Envie um arquivo (campo "file")');
    return this.demandasService.addAnexo(req.user.id, id, file, body?.nome);
  }

  @Get(':id/anexos/:anexoId/download')
  async downloadAnexo(
    @Req() req: { user: { id: string } },
    @Param('id') id: string,
    @Param('anexoId') anexoId: string,
    @Res() res: Response,
  ) {
    const download = await this.demandasService.getAnexoForDownload(req.user.id, id, anexoId);
    res.setHeader('Content-Type', download.mimeType);
    res.setHeader('Content-Disposition', `attachment; filename="${encodeURIComponent(download.filename)}"`);
    if ('buffer' in download) return res.send(download.buffer);
    return res.sendFile(download.path);
  }

  @Delete(':id/anexos/:anexoId')
  deleteAnexo(
    @Req() req: { user: { id: string } },
    @Param('id') id: string,
    @Param('anexoId') anexoId: string,
  ) {
    return this.demandasService.deleteAnexo(req.user.id, id, anexoId);
  }
}

File: backend-csharp-api/Controllers/DemandasController.cs
using LuxusDemandas.Api.Models;
using LuxusDemandas.Api.Security;
using LuxusDemandas.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace LuxusDemandas.Api.Controllers;

[ApiController]
[Authorize]
[Route("demandas")]
public sealed class DemandasController : ControllerBase
{
    private readonly DemandasService _demandas;

    public DemandasController(DemandasService demandas)
    {
        _demandas = demandas;
    }

    [HttpPost]
    public async Task<IActionResult> Create([FromBody] CreateDemandaRequest request, CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _demandas.CreateAsync(User.GetUserId(), request, cancellationToken));
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("from-template/{templateId}")]
    public async Task<IActionResult> CreateFromTemplate(string templateId, [FromBody] CreateDemandaFromTemplateRequest request, CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _demandas.CreateFromTemplateAsync(User.GetUserId(), templateId, request, cancellationToken));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
    }

    [HttpPost("buscar-ia")]
    public async Task<IActionResult> BuscarIa([FromBody] BuscarIaRequest request, CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _demandas.BuscarIaAsync(User.GetUserId(), request.Query, request.Scope, request.Context, cancellationToken));
        }
        catch (InvalidOperationException ex)
        {
            return StatusCode(StatusCodes.Status503ServiceUnavailable, new { message = ex.Message });
        }
    }

    [HttpGet("dashboard-kpis")]
    public async Task<IActionResult> DashboardKpis([FromQuery(Name = "avaliar_ia")] string? avaliarIa, CancellationToken cancellationToken)
    {
        if (!User.HasRoleSlug("admin"))
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = "Apenas usuário master (administrador) pode realizar esta ação." });
        }

        return Ok(await _demandas.GetDashboardKpisAsync(string.Equals(avaliarIa, "true", StringComparison.OrdinalIgnoreCase), cancellationToken));
    }

    [HttpGet]
    public async Task<IActionResult> List([FromQuery] ListDemandasFiltersQuery query, CancellationToken cancellationToken)
    {
        return Ok(await _demandas.ListAsync(User.GetUserId(), query, cancellationToken));
    }

    [HttpGet("export/excel")]
    public async Task<IActionResult> Export([FromQuery] ListDemandasFiltersQuery query, CancellationToken cancellationToken)
    {
        var data = await _demandas.ExportAsync(User.GetUserId(), query, cancellationToken);
        return Ok(data);
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> Detail(string id, CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _demandas.FindOneAsync(User.GetUserId(), id, cancellationToken));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
        catch (UnauthorizedAccessException ex)
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = ex.Message });
        }
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> Update(string id, [FromBody] UpdateDemandaRequest request, CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _demandas.UpdateAsync(User.GetUserId(), id, request, cancellationToken));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
        catch (UnauthorizedAccessException ex)
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> Remove(string id, CancellationToken cancellationToken)
    {
        if (!User.HasRoleSlug("admin"))
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = "Apenas usuário master (administrador) pode realizar esta ação." });
        }

        try
        {
            return Ok(await _demandas.RemoveAsync(User.GetUserId(), id, cancellationToken));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
        catch (UnauthorizedAccessException ex)
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = ex.Message });
        }
    }

    [HttpPost("{id}/observacoes")]
    public async Task<IActionResult> AddObservacao(string id, [FromBody] UpdateObservacaoRequest request, CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _demandas.AddObservacaoAsync(User.GetUserId(), id, request.Texto, cancellationToken));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
        catch (UnauthorizedAccessException ex)
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPatch("{id}/observacoes/{observacaoId}")]
    public async Task<IActionResult> UpdateObservacao(string id, string observacaoId, [FromBody] UpdateObservacaoRequest request, CancellationToken cancellationToken)
    {
        try
        {
            return Ok(await _demandas.UpdateObservacaoAsync(User.GetUserId(), id, observacaoId, request.Texto, cancellationToken));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
        catch (UnauthorizedAccessException ex)
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("{id}/anexos")]
    [RequestFormLimits(MultipartBodyLengthLimit = 100_000_000)]
    public async Task<IActionResult> AddAnexo(string id, [FromForm] IFormFile? file, [FromForm] string? nome, CancellationToken cancellationToken)
    {
        if (file is null || file.Length == 0)
        {
            return BadRequest(new { message = "Envie um arquivo (campo \"file\")" });
        }

        await using var stream = file.OpenReadStream();
        using var memory = new MemoryStream();
        await stream.CopyToAsync(memory, cancellationToken);

        try
        {
            return Ok(await _demandas.AddAnexoAsync(
                User.GetUserId(),
                id,
                memory.ToArray(),
                file.FileName,
                nome,
                file.ContentType,
                file.Length,
                cancellationToken));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
        catch (UnauthorizedAccessException ex)
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("{id}/anexos/{anexoId}/download")]
    public async Task<IActionResult> DownloadAnexo(string id, string anexoId, CancellationToken cancellationToken)
    {
        try
        {
            var download = await _demandas.GetAnexoForDownloadAsync(User.GetUserId(), id, anexoId, cancellationToken);
            return File(download.Buffer, download.MimeType, download.Filename);
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { message = ex.Message });
        }
        catch (UnauthorizedAccessException ex)
        {
            return StatusCode(StatusCodes.Status403Forbidden, new { message = ex.Message });
        }
    }
}
